Enable push-down of constant join conditions in cross and outer-joins

mkleen opened the pull request:
## Summary of the changes / Why this improves CrateDB

This extends `optimizer_move_constant_join_conditions_beneath_join` to also provide filter-push-down for cross and outer joins.

```
Q: select count(*) from doc.t1 right join doc.t2 on t1.id = t2.id and t1.id < 1000
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |     6180.172 ±  107.879 |   6037.348 |   6157.541 |   6255.526 |   6353.298 |
|   V2    |       77.206 ±   55.086 |     56.349 |     59.384 |     62.273 |    233.831 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               - 195.06%                           - 196.18%
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 195.06%
The test has statistical significance
```

## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

mkleen commented on server/src/test/java/io/crate/integrationtests/JoinIntegrationTest.java:
```diff
@@ -1335,15 +1335,15 @@ public void test_alias_in_left_join_condition() {
-                "    │  └ Collect[doc.t2 | [cluster_id] | (kind = 'bar')]",
-                "    └ Rename[cluster_id, kind] AS temp",
-                "      └ Collect[doc.t2 | [cluster_id, kind] | true]"
+            "Eval[id, reference]",
```
This is a nice example why this change makes sense. The constant join condition from the left-join gets extracted and pushed down and the join can be converted to inner-join which leads to a more efficient query plan using solely hash-joins.

```
+------------------------------------------------------+----------------------------------------------------------------------------------------+
| STEP                                                 | QUERY PLAN                                                                             |
+------------------------------------------------------+----------------------------------------------------------------------------------------+
| Initial logical plan                                 | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ Join[LEFT | ((cluster_id = id) AND (kind = 'bar'))] (rows=unknown)                 |
|                                                      |     ├ Join[INNER | ((cluster_id = id) AND (kind = 'bar'))] (rows=unknown)              |
|                                                      |     │  ├ Filter[(reference = 'bazinga')] (rows=0)                                      |
|                                                      |     │  │  └ Join[INNER | (subscription_id = id)] (rows=unknown)                        |
|                                                      |     │  │    ├ Collect[doc.t3 | [id, reference] | true] (rows=unknown)                  |
|                                                      |     │  │    └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)            |
|                                                      |     │  └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                    |
|                                                      |     └ Rename[cluster_id, kind] AS temp (rows=unknown)                                  |
|                                                      |       └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                     |
| optimizer_extract_constant_join_conditions_to_filter | Eval[id, reference] (rows=0)                                                           |
|                                                      |   └ Filter[(kind = 'bar')] (rows=0)                                                    |
|                                                      |     └ Join[LEFT | (cluster_id = id)] (rows=unknown)                                    |
|                                                      |       ├ Join[INNER | ((cluster_id = id) AND (kind = 'bar'))] (rows=unknown)            |
|                                                      |       │  ├ Filter[(reference = 'bazinga')] (rows=0)                                    |
|                                                      |       │  │  └ Join[INNER | (subscription_id = id)] (rows=unknown)                      |
|                                                      |       │  │    ├ Collect[doc.t3 | [id, reference] | true] (rows=unknown)                |
|                                                      |       │  │    └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)          |
|                                                      |       │  └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                  |
|                                                      |       └ Rename[cluster_id, kind] AS temp (rows=unknown)                                |
|                                                      |         └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                   |
| optimizer_rewrite_filter_on_outer_join_to_inner_join | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ Join[INNER | (cluster_id = id)] (rows=unknown)                                     |
|                                                      |     ├ Join[INNER | ((cluster_id = id) AND (kind = 'bar'))] (rows=unknown)              |
|                                                      |     │  ├ Filter[(reference = 'bazinga')] (rows=0)                                      |
|                                                      |     │  │  └ Join[INNER | (subscription_id = id)] (rows=unknown)                        |
|                                                      |     │  │    ├ Collect[doc.t3 | [id, reference] | true] (rows=unknown)                  |
|                                                      |     │  │    └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)            |
|                                                      |     │  └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                    |
|                                                      |     └ Filter[(kind = 'bar')] (rows=0)                                                  |
|                                                      |       └ Rename[cluster_id, kind] AS temp (rows=unknown)                                |
|                                                      |         └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                   |
| optimizer_rewrite_join_plan                          | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ HashJoin[(cluster_id = id)] (rows=unknown)                                         |
|                                                      |     ├ Join[INNER | ((cluster_id = id) AND (kind = 'bar'))] (rows=unknown)              |
|                                                      |     │  ├ Filter[(reference = 'bazinga')] (rows=0)                                      |
|                                                      |     │  │  └ Join[INNER | (subscription_id = id)] (rows=unknown)                        |
|                                                      |     │  │    ├ Collect[doc.t3 | [id, reference] | true] (rows=unknown)                  |
|                                                      |     │  │    └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)            |
|                                                      |     │  └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                    |
|                                                      |     └ Filter[(kind = 'bar')] (rows=0)                                                  |
|                                                      |       └ Rename[cluster_id, kind] AS temp (rows=unknown)                                |
|                                                      |         └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                   |
| optimizer_extract_constant_join_conditions_to_filter | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ HashJoin[(cluster_id = id)] (rows=unknown)                                         |
|                                                      |     ├ Filter[(kind = 'bar')] (rows=0)                                                  |
|                                                      |     │  └ Join[INNER | (cluster_id = id)] (rows=unknown)                                |
|                                                      |     │    ├ Filter[(reference = 'bazinga')] (rows=0)                                    |
|                                                      |     │    │  └ Join[INNER | (subscription_id = id)] (rows=unknown)                      |
|                                                      |     │    │    ├ Collect[doc.t3 | [id, reference] | true] (rows=unknown)                |
|                                                      |     │    │    └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)          |
|                                                      |     │    └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                  |
|                                                      |     └ Filter[(kind = 'bar')] (rows=0)                                                  |
|                                                      |       └ Rename[cluster_id, kind] AS temp (rows=unknown)                                |
|                                                      |         └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                   |
| optimizer_move_filter_beneath_join                   | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ HashJoin[(cluster_id = id)] (rows=unknown)                                         |
|                                                      |     ├ Join[INNER | (cluster_id = id)] (rows=unknown)                                   |
|                                                      |     │  ├ Filter[(reference = 'bazinga')] (rows=0)                                      |
|                                                      |     │  │  └ Join[INNER | (subscription_id = id)] (rows=unknown)                        |
|                                                      |     │  │    ├ Collect[doc.t3 | [id, reference] | true] (rows=unknown)                  |
|                                                      |     │  │    └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)            |
|                                                      |     │  └ Filter[(kind = 'bar')] (rows=0)                                               |
|                                                      |     │    └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                  |
|                                                      |     └ Filter[(kind = 'bar')] (rows=0)                                                  |
|                                                      |       └ Rename[cluster_id, kind] AS temp (rows=unknown)                                |
|                                                      |         └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                   |
| optimizer_rewrite_join_plan                          | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ HashJoin[(cluster_id = id)] (rows=unknown)                                         |
|                                                      |     ├ HashJoin[(cluster_id = id)] (rows=unknown)                                       |
|                                                      |     │  ├ Filter[(reference = 'bazinga')] (rows=0)                                      |
|                                                      |     │  │  └ Join[INNER | (subscription_id = id)] (rows=unknown)                        |
|                                                      |     │  │    ├ Collect[doc.t3 | [id, reference] | true] (rows=unknown)                  |
|                                                      |     │  │    └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)            |
|                                                      |     │  └ Filter[(kind = 'bar')] (rows=0)                                               |
|                                                      |     │    └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                  |
|                                                      |     └ Filter[(kind = 'bar')] (rows=0)                                                  |
|                                                      |       └ Rename[cluster_id, kind] AS temp (rows=unknown)                                |
|                                                      |         └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                   |
| optimizer_move_filter_beneath_join                   | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ HashJoin[(cluster_id = id)] (rows=unknown)                                         |
|                                                      |     ├ HashJoin[(cluster_id = id)] (rows=unknown)                                       |
|                                                      |     │  ├ Join[INNER | (subscription_id = id)] (rows=unknown)                           |
|                                                      |     │  │  ├ Filter[(reference = 'bazinga')] (rows=0)                                   |
|                                                      |     │  │  │  └ Collect[doc.t3 | [id, reference] | true] (rows=unknown)                 |
|                                                      |     │  │  └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)              |
|                                                      |     │  └ Filter[(kind = 'bar')] (rows=0)                                               |
|                                                      |     │    └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                  |
|                                                      |     └ Filter[(kind = 'bar')] (rows=0)                                                  |
|                                                      |       └ Rename[cluster_id, kind] AS temp (rows=unknown)                                |
|                                                      |         └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                   |
| optimizer_rewrite_join_plan                          | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ HashJoin[(cluster_id = id)] (rows=unknown)                                         |
|                                                      |     ├ HashJoin[(cluster_id = id)] (rows=unknown)                                       |
|                                                      |     │  ├ HashJoin[(subscription_id = id)] (rows=unknown)                               |
|                                                      |     │  │  ├ Filter[(reference = 'bazinga')] (rows=0)                                   |
|                                                      |     │  │  │  └ Collect[doc.t3 | [id, reference] | true] (rows=unknown)                 |
|                                                      |     │  │  └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)              |
|                                                      |     │  └ Filter[(kind = 'bar')] (rows=0)                                               |
|                                                      |     │    └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                  |
|                                                      |     └ Filter[(kind = 'bar')] (rows=0)                                                  |
|                                                      |       └ Rename[cluster_id, kind] AS temp (rows=unknown)                                |
|                                                      |         └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                   |
| optimizer_merge_filter_and_collect                   | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ HashJoin[(cluster_id = id)] (rows=unknown)                                         |
|                                                      |     ├ HashJoin[(cluster_id = id)] (rows=unknown)                                       |
|                                                      |     │  ├ HashJoin[(subscription_id = id)] (rows=unknown)                               |
|                                                      |     │  │  ├ Collect[doc.t3 | [id, reference] | (reference = 'bazinga')] (rows=unknown) |
|                                                      |     │  │  └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)              |
|                                                      |     │  └ Filter[(kind = 'bar')] (rows=0)                                               |
|                                                      |     │    └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                  |
|                                                      |     └ Filter[(kind = 'bar')] (rows=0)                                                  |
|                                                      |       └ Rename[cluster_id, kind] AS temp (rows=unknown)                                |
|                                                      |         └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                   |
| optimizer_merge_filter_and_collect                   | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ HashJoin[(cluster_id = id)] (rows=unknown)                                         |
|                                                      |     ├ HashJoin[(cluster_id = id)] (rows=unknown)                                       |
|                                                      |     │  ├ HashJoin[(subscription_id = id)] (rows=unknown)                               |
|                                                      |     │  │  ├ Collect[doc.t3 | [id, reference] | (reference = 'bazinga')] (rows=unknown) |
|                                                      |     │  │  └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)              |
|                                                      |     │  └ Collect[doc.t2 | [cluster_id, kind] | (kind = 'bar')] (rows=unknown)          |
|                                                      |     └ Filter[(kind = 'bar')] (rows=0)                                                  |
|                                                      |       └ Rename[cluster_id, kind] AS temp (rows=unknown)                                |
|                                                      |         └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                   |
| optimizer_move_filter_beneath_rename                 | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ HashJoin[(cluster_id = id)] (rows=unknown)                                         |
|                                                      |     ├ HashJoin[(cluster_id = id)] (rows=unknown)                                       |
|                                                      |     │  ├ HashJoin[(subscription_id = id)] (rows=unknown)                               |
|                                                      |     │  │  ├ Collect[doc.t3 | [id, reference] | (reference = 'bazinga')] (rows=unknown) |
|                                                      |     │  │  └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)              |
|                                                      |     │  └ Collect[doc.t2 | [cluster_id, kind] | (kind = 'bar')] (rows=unknown)          |
|                                                      |     └ Rename[cluster_id, kind] AS temp (rows=0)                                        |
|                                                      |       └ Filter[(kind = 'bar')] (rows=0)                                                |
|                                                      |         └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                   |
| optimizer_merge_filter_and_collect                   | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ HashJoin[(cluster_id = id)] (rows=unknown)                                         |
|                                                      |     ├ HashJoin[(cluster_id = id)] (rows=unknown)                                       |
|                                                      |     │  ├ HashJoin[(subscription_id = id)] (rows=unknown)                               |
|                                                      |     │  │  ├ Collect[doc.t3 | [id, reference] | (reference = 'bazinga')] (rows=unknown) |
|                                                      |     │  │  └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)              |
|                                                      |     │  └ Collect[doc.t2 | [cluster_id, kind] | (kind = 'bar')] (rows=unknown)          |
|                                                      |     └ Rename[cluster_id, kind] AS temp (rows=unknown)                                  |
|                                                      |       └ Collect[doc.t2 | [cluster_id, kind] | (kind = 'bar')] (rows=unknown)           |
| Final logical plan                                   | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ HashJoin[(cluster_id = id)] (rows=unknown)                                         |
|                                                      |     ├ HashJoin[(cluster_id = id)] (rows=unknown)                                       |
|                                                      |     │  ├ HashJoin[(subscription_id = id)] (rows=unknown)                               |
|                                                      |     │  │  ├ Collect[doc.t3 | [id, reference] | (reference = 'bazinga')] (rows=unknown) |
|                                                      |     │  │  └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)              |
|                                                      |     │  └ Collect[doc.t2 | [cluster_id] | (kind = 'bar')] (rows=unknown)                |
|                                                      |     └ Rename[cluster_id] AS temp (rows=unknown)                                        |
|                                                      |       └ Collect[doc.t2 | [cluster_id] | (kind = 'bar')] (rows=unknown)                 |
+------------------------------------------------------+----------------------------------------------------------------------------------------+
EXPLAIN 14 rows in set (0.026 sec)
```

mkleen commented on server/src/main/java/io/crate/planner/optimizer/rule/ExtractConstantJoinConditionsToFilter.java:
```diff
@@ -88,21 +96,17 @@ public LogicalPlan apply(JoinPlan joinPlan,
-                joinPlan.isFiltered(),
-                joinPlan.isRewriteFilterOnOuterJoinToInnerJoinDone(),
-                joinPlan.isLookUpJoinRuleApplied()
+            return new Filter(
+                new JoinPlan(
+                    joinPlan.lhs(),
+                    joinPlan.rhs(),
+                    joinPlan.joinType(),
+                    AndOperator.join(nonConstantConditions),
+                    joinPlan.isFiltered(),
+                    joinPlan.isRewriteFilterOnOuterJoinToInnerJoinDone(),
+                    joinPlan.isLookUpJoinRuleApplied(),
+                    true),
+                AndOperator.join(constantConditions.values())
```
If the join creates extra null rows like left/right `MoveFilterBeneathJoin` will only push the condition to the side which does not produce null values.

mkleen replied:
So in case the constant join condition references a side which produces null rows, the join condition will become a filter and stay there.

mkleen replied:
Maybe its worth benchmarking this case.

mkleen replied:
This case won't happen anymore.

mkleen commented:
@mfussenegger I made a second iteration to handle outer joins properly.  Outer-joins only apply constant join conditions on their non-preserved side:

- LEFT JOIN will only filter on the RHS
- RIGHT JOIN will only filter on the LHS
-  
 Therefore we can only extract the filter in these cases on top of the join

mkleen commented:
I think the real benefit of this optimization concept is that it converts non-equi-join to equi-joins which then leads to further optimizations. With having https://github.com/crate/crate/issues/15147 in mind, i think this is the right direction.

mkleen commented on server/src/test/java/io/crate/integrationtests/TransportSQLActionTest.java:
```diff
@@ -1697,6 +1697,7 @@ public void test_select_from_virtual_table_with_window_function_and_column_pruni
         // push down of columns used in WHERE on top of join results in column pruning
         // This triggered a bug in the prune implementation of the WindowAgg operator
         execute("create table doc.metrics (id int primary key, x int)");
+        waitNoPendingTasksOnAll();
```
The test is flaky, this fixes it.

mkleen commented:
The test is still flaky, i will put this on hold.

mkleen commented on server/src/main/java/io/crate/planner/operators/LogicalPlanner.java:
```diff
@@ -147,6 +147,7 @@ public class LogicalPlanner {
         new RemoveRedundantEval(),
         new MergeAggregateAndCollectToCount(),
         new MergeAggregateRenameAndCollectToCount(),
+        new ExtractConstantJoinConditionsToFilter(),
```
Let's have this rule before the other filter push-down rules apply.

matriv commented on docs/appendices/release-notes/5.8.0.rst:
```diff
@@ -116,6 +116,19 @@ Scalar and Aggregation Functions
+
+    select * from t1 left join t2 on t1.id = t2.id and t1.id > 1
+
+  This optimization can be disabled if desired, with the session setting::
```
minor:
```suggestion
  This optimization can be disabled, if desired, with the session setting::
```

matriv commented on server/src/main/java/io/crate/planner/optimizer/rule/ExtractConstantJoinConditionsToFilter.java:
```diff
@@ -0,0 +1,155 @@
+
+/**
+ * This rule extracts constant join conditions from inner/outer/cross joins and places them as
+ * a filter on top of the join. Then other rules can process the filter further.
```
Maybe it's too much, just a suggestion:
```suggestion
 * a filter on top of the join. Then other rules can process the filter further,
 * for example, push it beneath the join to a collect phase.
```

matriv commented on server/src/test/java/io/crate/planner/operators/JoinTest.java:
```diff
@@ -1098,4 +1098,24 @@ public void test_relationnames_order_in_nested_loop_join() throws Exception {
     }
+
+    @Test
+    public void test_push_constant_join_conditions_beyond_outer_joins() throws Exception {
```
```suggestion
    public void test_push_constant_join_conditions_beneath_outer_joins() throws Exception {
```

matriv commented on server/src/main/java/io/crate/planner/optimizer/rule/ExtractConstantJoinConditionsToFilter.java:
```diff
@@ -0,0 +1,155 @@
+            );
+    }
+
+    private static boolean hasConstantJoinConditions(Symbol joinCondition) {
```
I'd remove this method and the check in the pattern matching, to avoid calling the `QuerySplitter.split()` twice. I think it's fine to just bail out early in `apply()` or "cache" the outcome of the splitter in a variable.

mkleen replied:
This is on purpose in the pattern matching, to make sure this rule only matches if there is a constant join condition present. If i put this into the apply it will show up on `explain verbose` without doing anything.

matriv replied:
Makes sense, thx for explaining! Consider if there is a solution to not do the query splitting twice then.

matriv commented on server/src/main/java/io/crate/planner/optimizer/rule/ExtractConstantJoinConditionsToFilter.java:
```diff
@@ -0,0 +1,155 @@
+        for (var condition : allConditions.entrySet()) {
+            Set<RelationName> relationsInCurrentConditions = RelationNameCollector.collect(condition.getValue());
+            if (relationsInCurrentConditions.size() == 1) {
+                // Constant join conditions can only pushed to the preserved side of a join, where no null values are created:
```
```suggestion
                // Constant join conditions can only be pushed to the preserved side of a join, where no null values are created:
```

matriv commented on server/src/test/java/io/crate/planner/operators/JoinTest.java:
```diff
@@ -1098,4 +1098,24 @@ public void test_relationnames_order_in_nested_loop_join() throws Exception {
+        QueriedSelectRelation mss = e.analyze("SELECT * FROM t1 LEFT JOIN t2 ON t1.a = t2.b AND t1.a > 1");
+
+        var plannerCtx = executor.getPlannerContext();
+        plannerCtx.transactionContext().sessionSettings().setHashJoinEnabled(false);
```
this is not needed, and I think it can be confusing, as currently outer joins are executed only via NestedLoop.

matriv commented on server/src/main/java/io/crate/planner/optimizer/rule/ExtractConstantJoinConditionsToFilter.java:
```diff
@@ -0,0 +1,155 @@
+                remainingConditions.add(condition.getValue());
+            }
+        }
+        if (!constantConditionsForFilter.isEmpty() && !remainingConditions.isEmpty()) {
```
I think the `!remainingConditions.isEmpty()` should be removed, e.g:
```
SELECT * FROM t1 JOIN t2 ON t1.a > 10 AND t1.a > 1
```
here we should extract both as a filter which will be pushed down to t1 afterwards.
But maybe it needs further processing, as the JOIN currently will remain `INNER` where essentially it should become `CROSS`. I think the same goes for LEFT/RIGHT, they become `CROSS`.

It can definitely be addressed as follow-up, since it's an edge case where there are no conditions involving cols from the 2 tables.

matriv commented on server/src/main/java/io/crate/planner/optimizer/rule/ExtractConstantJoinConditionsToFilter.java:
```diff
@@ -0,0 +1,155 @@
+                if (joinPlan.joinType().isOuter()) {
+                    var matchesLhs = Sets.intersection(lhsRelations, relationsInCurrentConditions);
+                    var matchesRhs = Sets.intersection(rhsRelations, relationsInCurrentConditions);
+                    if ((!matchesLhs.isEmpty() && joinPlan.joinType() == LEFT) ||
```
I compared to postgres and it seems that this check should be inverted.
I still don't get it, it seems counter intuitive, but check it out:
psql:
```
matriv=> create table t1(a int);
CREATE TABLE
matriv=> create table t2(b int);
CREATE TABLE
matriv=> insert into t1 select * from generate_series(1, 10, 1);
INSERT 0 10
matriv=> insert into t2 select * from generate_series(6, 10, 1);
INSERT 0 5

matriv=> SELECT * FROM t1 LEFT JOIN t2 ON t1.a > 5 AND t1.a = t2.b;
 a  | b  
----+----
  1 |   
  2 |   
  3 |   
  4 |   
  5 |   
  6 |  6
  7 |  7
  8 |  8
  9 |  9
 10 | 10
(10 rows)
matriv=> explain SELECT * FROM t1 LEFT JOIN t2 ON t1.a > 5 AND t1.a = t2.b;
                            QUERY PLAN                            
------------------------------------------------------------------
 Merge Left Join  (cost=359.57..941.28 rows=10838 width=8)
   Merge Cond: (t1.a = t2.b)
   Join Filter: (t1.a > 5)
   ->  Sort  (cost=179.78..186.16 rows=2550 width=4)
         Sort Key: t1.a
         ->  Seq Scan on t1  (cost=0.00..35.50 rows=2550 width=4)
   ->  Sort  (cost=179.78..186.16 rows=2550 width=4)
         Sort Key: t2.b
         ->  Seq Scan on t2  (cost=0.00..35.50 rows=2550 width=4)
(9 rows)

matriv=> SELECT * FROM t1 RIGHT JOIN t2 ON t1.a > 5 AND t1.a = t2.b;
 a  | b  
----+----
  6 |  6
  7 |  7
  8 |  8
  9 |  9
 10 | 10
(5 rows)
matriv=> explain SELECT * FROM t1 RIGHT JOIN t2 ON t1.a > 5 AND t1.a = t2.b;
                            QUERY PLAN                            
------------------------------------------------------------------
 Merge Right Join  (cost=263.02..429.84 rows=10838 width=8)
   Merge Cond: (t1.a = t2.b)
   ->  Sort  (cost=83.23..85.36 rows=850 width=4)
         Sort Key: t1.a
         ->  Seq Scan on t1  (cost=0.00..41.88 rows=850 width=4)
               Filter: (a > 5)
   ->  Sort  (cost=179.78..186.16 rows=2550 width=4)
         Sort Key: t2.b
         ->  Seq Scan on t2  (cost=0.00..35.50 rows=2550 width=4)
(9 rows)
```

crate (this PR):
```
cr> SELECT * FROM t1 LEFT JOIN t2 ON t1.a > 5 AND t1.a = t2.b;
+----+----+
|  a |  b |
+----+----+
|  7 |  7 |
|  8 |  8 |
|  9 |  9 |
| 10 | 10 |
|  6 |  6 |
+----+----+
SELECT 5 rows in set (1.223 sec)
cr> SELECT * FROM t1 RIGHT JOIN t2 ON t1.a > 5 AND t1.a = t2.b;
+----+----+
|  a |  b |
+----+----+
|  7 |  7 |
|  8 |  8 |
|  9 |  9 |
| 10 | 10 |
|  6 |  6 |
+----+----+
SELECT 5 rows in set (0.988 sec)
```


checkout how psql produces different result if the constant is moved from `ON` to `WHERE`:
```
matriv=> SELECT * FROM t1 LEFT JOIN t2 ON t1.a > 5 AND t1.a = t2.b;
 a  | b  
----+----
  1 |   
  2 |   
  3 |   
  4 |   
  5 |   
  6 |  6
  7 |  7
  8 |  8
  9 |  9
 10 | 10
(10 rows)
                                        ^
matriv=> SELECT * FROM t1 LEFT JOIN t2 ON t1.a = t2.b WHERE t1.a > 5;
 a  | b  
----+----
  6 |  6
  7 |  7
  8 |  8
  9 |  9
 10 | 10
(5 rows)

```

Can you please check some other database?
Can you also run the `crate-qa` sqllogictests against your branch to see if it breaks something?

matriv replied:
To demonstrate it better (filter is removing more rows than the matches between t1 and t2):
```
matriv=> SELECT * FROM t1 LEFT JOIN t2 ON t1.a > 8 AND t1.a = t2.b;
 a  | b  
----+----
  1 |   
  2 |   
  3 |   
  4 |   
  5 |   
  6 |   
  7 |   
  8 |   
  9 |  9
 10 | 10
(10 rows)

matriv=> SELECT * FROM t1 LEFT JOIN t2 ON t1.a = t2.b WHERE t1.a > 8;
 a  | b  
----+----
  9 |  9
 10 | 10
(2 rows)

```

matriv reviewed (commented):
Thx for working on this to further improve our join performance.
I've left some minor wording comments, but I found something weird when testing against PostgreSQL, check the corresponding comment on the rule please.

mkleen committed:
Enable push-down of constant join conditions in cross and outer-joins

This extends optimizer_move_constant_join_conditions_beneath_join to also
provide filter push down for cross and left/right joins.

mkleen commented:
@matriv Different approach with the same behaviour. This is 100 % save to do so. 

```
cr> SELECT * FROM t1 LEFT JOIN t2 ON t1.a > 5 AND t1.a = t2.b;
+----+------+
|  a |    b |
+----+------+
|  3 | NULL |
|  6 |    6 |
|  8 |    8 |
|  9 |    9 |
|  2 | NULL |
|  5 | NULL |
|  1 | NULL |
|  7 |    7 |
| 10 |   10 |
|  4 | NULL |
+----+------+


cr> SELECT * FROM t1 RIGHT JOIN t2 ON t1.a > 5 AND t1.a = t2.b;
+----+----+
|  a |  b |
+----+----+
|  6 |  6 |
|  8 |  8 |
|  9 |  9 |
|  7 |  7 |
| 10 | 10 |
+----+----+
```

mkleen commented:
Will now also run the sql-logic tests on this.

matriv reviewed (approved):
thx @mkleen!

Left a minor comment.
If it's not too much trouble maybe you can run and post a representative benchmark that shows the perf improvement we get.

mkleen commented on server/src/main/java/io/crate/planner/optimizer/rule/MoveConstantJoinConditionsBeneathJoin.java:
```diff
@@ -88,25 +99,54 @@ public LogicalPlan apply(JoinPlan joinPlan,
+                return joinPlan.withMoveConstantJoinConditionRuleApplied(true);
+            }
+            var newJoinConditions = new LinkedHashSet<>(nonConstantConditions);
+            if (!constantConditions.isEmpty()) {
```
good point.

mkleen commented:
> thx @mkleen!
> 
> Left a minor comment. If it's not too much trouble maybe you can run and post a representative benchmark that shows the perf improvement we get.

This depends really on scenario and how much the filter will reduce the number of join operation to process by pushing it down to the collect. The impact of a filter push-down can be really significant e.g.:

```
Q: select count(*) from doc.t1 right join doc.t2 on t1.id = t2.id and t1.id < 1000
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |     6180.172 ±  107.879 |   6037.348 |   6157.541 |   6255.526 |   6353.298 |
|   V2    |       77.206 ±   55.086 |     56.349 |     59.384 |     62.273 |    233.831 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               - 195.06%                           - 196.18%
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 195.06%
The test has statistical significance
```

matriv commented:
> Note: This is a benchmark where it works well, created on purpose for this use-case.

Yep, of course benchmarks depend on the exact query and data, but this one can show well how much this enchanced rule can help for such cases.
